bin/packages/check-build-type-declaration-files.js

```diff
@@ -0,0 +1,114 @@
+/**
+ * The purpose of this script is to double-check certain published build types.
+ * When we translate JSDoc into types that we wish to publish, we occasionally
+ * disable the type checker for JS files. Unfortunately, this can result in invalid
+ * index.d.ts files being published.
+ *
+ * This script verifies the publishd index.d.ts file for every package which both
+ * builds types and also sets checkJs to false in its tsconfig.json. This is done
+ * by running `tsc --noEmit` on the $package/build-types/index.d.ts file. This also
+ * verifies everything index.d.ts references, which essentially checks the entire
+ * public api of the type declarations for that package.
+ *
+ * @see https://github.com/WordPress/gutenberg/pull/49650 for more discussion.
+ */
+
+/**
+ * External dependencies
+ */
+const fs = require( 'fs' ).promises;
+const path = require( 'path' );
+const { exec } = require( 'child_process' );
+const chalk = require( 'chalk' );
+
+// Tsconfig can be JSON5, which accepts comments, which need to be stripped so we can parse the JSON.
+function stripJsonComments( jsonString ) {
+	const lines = jsonString.split( '\n' );
+	const cleanedLines = lines.map( ( line ) => line.replace( /\/\/.*$/, '' ) );
+	return cleanedLines.join( '\n' );
+}
+
+async function readTsConfig( tsconfigPath ) {
+	const tsconfigRaw = await fs.readFile( tsconfigPath, 'utf-8' );
+	return JSON.parse( stripJsonComments( tsconfigRaw ) );
+}
+
+async function typecheckDeclarations( file ) {
+	return new Promise( ( resolve, reject ) => {
+		exec( `npx tsc --noEmit ${ file }`, ( error, stdout, stderr ) => {
+			if ( error ) {
+				reject( { file, error, stderr, stdout } );
+			} else {
+				resolve( { file, stdout } );
+			}
+		} );
+	} );
+}
+
+// Returns the path to the build-types declaration file for a package if it exists.
+// Throws an error and exits the script otherwise.
+async function getDecFile( packagePath ) {
+	const decFile = path.join( packagePath, 'build-types', 'index.d.ts' );
+	try {
+		await fs.access( decFile );
+		return decFile;
+	} catch ( err ) {
+		if ( err.code !== 'ENOENT' ) {
+			throw err;
+		}
+		console.error(
+			`This declaration file should exist. You may need to run tsc again: ${ decFile }`
+		);
+		process.exit( 1 );
+	}
+}
```

Comment:
> in this function I think fileExistsSync is probably still appropriate since our purpose is to check the presence of the file but not read it in.

I'm curious about the difference between your approach vs `fs.access`. It seems `fs.access` is the right tool for checking if we have permissions to access the file (and I've definitely seen it recommended for this kind of scenario elsewhere). In this case, we don't need to read the declaration file; we just need to make sure it exists. 🤔 

> given that readTsConfig is only used in determining if checkJs is true, we could inline that as well as the error handling

I refactored to include some of these changes, moving more of the logic outside of the main function.

Reply:
> It seems fs.access is the right tool for checking if we have permissions to access the file (and I've definitely seen it recommended for this kind of scenario elsewhere). In this case, we don't need to read the declaration file; we just need to make sure it exists.

I might have missed that we never read the declarations file, I know at some point I was confused at least between the declarations file and the `tsconfig.json` file.

Didn't mean to say `access` or `fileExistsSync` was wrong, just that _in the case where we want to give a first pass checking for the presence of the file_ and then later read that file `fs.open()` provides a convenient wway to do that without introducing the race-condition (a condition I see holding low-risk here anyway because we're not looking at running this script when those files are likely to change)

Reply:
Cool!